README.md
# spark cookbook

A very basic resource for adding spark to a machine. Currently only supports standalone spark and does not attempt to manage any daemons or running spark processes.

## Supported Platforms

  * Debian/Ubuntu

## Usage

### spark_standalone resource

Include a `spark_standalone` resource in your cookbook to add the spark binary and resources to your system:

```ruby
spark_standalone 'default' do # note: resource name is not used in this resource
  # spark_version '1.6.2'
  # hadoop_version '2.6' 
  # base_path '/usr/local/spark'
  # log_level 'INFO'
end
```

Attribute details:
  
  * `spark_version` and `hadoop_version` are used directly in downloading a tarball (eg. https://d3kbcqa49mib13.cloudfront.net/spark-1.6.2-bin-hadoop2.6.tgz)
  * `hadoop_version` will default to `2.6` when a spark version < 2.0.0 is used. `2.7` otherwise.
  * `base_path` downloaded versions will be placd into this directory based on their version name.
  * `log_level` the log4j properties file will be customized to include this default log level.

The included `standalone` recipe provides some simple examples of usage, and a quick accessory to get some common sparks included.

### spark::standalone recipe

Include `spark::standalone` in your node's `run_list` to have your system configured with a java as well as spark 1.6.2 and 2.0.0:

```json
{
  "run_list": [
    "recipe[spark::standalone]"
  ]
}
```

You'll now have spark available from the locations below. set `SPARK_HOME` and give it a try.
  * `/usr/local/spark/spark-1.6.2-bin-hadoop2.6/`
  * `/usr/local/spark/spark-2.0.0-bin-hadoop2.7/`

## License and Authors

Author:: Steve Nolen (<steve.nolen@rstudio.com>)
Copyright:: 2016, RStudio Inc. 

```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# 0.1.0

Initial release of spark


libraries/provider_spark_standalone.rb
require 'chef/provider/lwrp_base'
require_relative 'helpers'

class Chef
  class Provider
    class SparkStandalone < Chef::Provider::LWRPBase
      use_inline_resources if defined?(use_inline_resources)

      def whyrun_supported?
        true
      end

      # Mix in helpers from libraries/helpers.rb
      include SparkCookbook::Helpers

      action :create do
        # if user supplies hadoop version, use it explicitly. if not, case depending on spark version selected
        hadoop_version =
          if new_resource.hadoop_version.nil?
            get_default_hadoop(new_resource.spark_version)
          else
            new_resource.hadoop_version
          end
        version_name = "spark-#{new_resource.spark_version}-bin-hadoop#{hadoop_version}"
        path = ::File.join(new_resource.base_path, version_name) + '/'

        directory path do
          recursive true
        end

        remote_file "spark_standalone #{new_resource.instance}: download spark tarball" do
          path "#{Chef::Config[:file_cache_path]}/#{version_name}.tar.gz"
          source "https://d3kbcqa49mib13.cloudfront.net/#{version_name}.tgz"
          notifies :run, "bash[spark_standalone #{new_resource.instance}: extract spark tarball]", :immediately
        end

        bash "spark_standalone #{new_resource.instance}: extract spark tarball" do
          cwd Chef::Config[:file_cache_path]
          code "tar xf #{version_name}.tar.gz -C #{path} --strip-components=1"
          action :nothing
        end

        template "spark_standalone #{new_resource.instance}: create hive-site.xml" do
          cookbook 'spark'
          path "#{path}conf/hive-site.xml"
          source 'hive-site.xml.erb'
          variables(resource: new_resource)
        end

        template "spark_standalone #{new_resource.instance}: create log4j.properties" do
          cookbook 'spark'
          path "#{path}conf/log4j.properties"
          source 'log4j.properties.erb'
          variables(resource: new_resource)
        end
      end
    end
  end
end


libraries/spark_standalone.rb
module SparkCookbook
  class SparkStandalone < ChefCompat::Resource
    use_automatic_resource_name
    provides :spark_standalone

    property :instance, String, name_attribute: true
    property :spark_version, String, default: '1.6.2'
    property :base_path, String, default: '/usr/local/spark' # sparks will be created within this directory
    property :log_level, String, default: 'INFO'
    property :hadoop_version, String

    ################
    # Helper Methods
    ################
    def get_default_hadoop(spark_version)
      return '2.7' if Gem::Version.new(spark_version) >= Gem::Version.new('2.0')
      '2.6'
    end

    ################
    # create method
    ################
    action :create do
      # if user supplies hadoop version, use it explicitly. if not, case depending on spark version selected
      hadoop_version = property_is_set?(:hadoop_version) ? new_resource.hadoop_version : get_default_hadoop(new_resource.spark_version)

      version_name = "spark-#{new_resource.spark_version}-bin-hadoop#{hadoop_version}"
      path = ::File.join(new_resource.base_path, version_name) + '/'

      directory path do
        recursive true
      end

      remote_file "spark_standalone #{new_resource.instance}: download spark tarball" do
        path "#{Chef::Config[:file_cache_path]}/#{version_name}.tar.gz"
        source "https://d3kbcqa49mib13.cloudfront.net/#{version_name}.tgz"
        notifies :run, "bash[spark_standalone #{new_resource.instance}: extract spark tarball]", :immediately
      end

      bash "spark_standalone #{new_resource.instance}: extract spark tarball" do
        cwd Chef::Config[:file_cache_path]
        code "tar xf #{version_name}.tar.gz -C #{path} --strip-components=1"
        action :nothing
      end

      template "spark_standalone #{new_resource.instance}: create hive-site.xml" do
        cookbook 'spark'
        path "#{path}conf/hive-site.xml"
        source 'hive-site.xml.erb'
        variables(resource: new_resource)
      end

      template "spark_standalone #{new_resource.instance}: create log4j.properties" do
        cookbook 'spark'
        path "#{path}conf/log4j.properties"
        source 'log4j.properties.erb'
        variables(resource: new_resource)
      end
    end
  end
end
